src/webpack/index.ts

```diff
@@ -75,3 +84,55 @@ function removeMiniCssExtract(config: any) {
     }
   });
 }
+
+function getLibraryName(options: any) {
+  if (options?.customWebpackConfig?.libraryName) {
+    return options.customWebpackConfig.libraryName;
+  }
+
+  const projectName = getProjectNameFromAngularJson();
+  if (projectName) return projectName;
+
+  console.warn(
+    'Warning: single-spa-angular could not determine a library name to use and has used a default value.',
+  );
+  console.info('This may cause issues if this app uses code-splitting or lazy loading.');
+  if (!options) {
+    console.info('You may also need to update extra-webpack.config.json.');
+  }
+  console.info('See <SomeUrl> for information on how to resolve this.');
+
+  return 'angular_single_spa_project';
+}
+
+function getProjectNameFromAngularJson(): string | null {
+  const angularJsonPath = findUp(['angular.json', '.angular.json']);
+  if (!angularJsonPath) return null;
+
+  // eslint-disable-next-line @typescript-eslint/no-var-requires
+  const angularJson = require(angularJsonPath);
+  if (!angularJson.projects) return null;
+
+  var projects = Object.getOwnPropertyNames(angularJson.projects);
+
+  if (projects.length === 1) return projects[0];
+
+  // If we reach here there are multiple (or zero) projects in angular.json
+  // we cannot tell which one to use, so we will end up using the default.
+  return null;
+}
+
+function findUp(files: string[]): string | null {
```

Comment:
From my subjective perspective, this seems like a complicated implementation.

We have to be using tools that Angular DevKit already provides and you can work with Angular config using native tools:
```js
const { workspaces } = require('@angular-devkit/core');
const { NodeJsSyncHost } = require('@angular-devkit/core/node');

const host = workspaces.createWorkspaceHost(new NodeJsSyncHost());

workspaces.readWorkspace(`${process.cwd()}/angular.json`, host).then(({ workspace }) => {
  console.log(workspace.projects.keys());
});
```

It's also worth adding comments to the code. I've read it multiple times and I can't understand what's going on in `while + nested for` :smile:
Excuse my pedantry :)

Reply:
Yeah, the while loop is broken. I must have accidentally lost a line.

The problem I am trying to tackle here is that there is zero guarantee that angular.json is in the same folder as `process.cwd()`. This can happen if you run `ng build` from inside a source subfolder.
In order to find the workspace folder, upward searching is required.

All that is required before you can even call `.readWorkspace()`. 

angular-cli does its equivalent findUp logic here: https://github.com/angular/angular-cli/blob/0070ea46e6eaa25ad49d17c24fe7bed546dc1ec0/packages/angular/cli/utilities/project.ts#L29.

Once we have located the file, then yeah, we *could* use the devkit workspaces, except this is 100% synchronous code, and dev-kit workspaces is async.

Reply:
Webpack config can export a Promise or a factory that returns Promise thus you're able to work with sync/async code.

Reply:
Yeah, but that defeats the whole point of the angular.json parsing, which is as a fallback that just works for the majority of all users if they are upgrading from an older version of the library. 

If the users need to update the `extra-webpack.config.ts` to make it export an promise returning function, then they could instead update it to pass the options argument to `singleSpaAngularWebpack` again, and update angular.json to specify the new 'libraryName' option. The new option (automatically populated by the schematic on new projects) is the primary source of the library name, with the angular.json parsing fallback as secondary, and a static value + warning as the third.

We could skip all angular.json parsing, but if we did so, then all existing users would either start getting the warning about using the default, and would need to make changes to silence it. The current approach is that only those few users who have multi-project workspaces will get the warning in the first place, as for everybody else we can infer it from the one-and-only project in angular.json.

Reply:
Okay. Can't we re-use the existing function then and not copy it from their source code?:
```js
const { findUp } = require('@angular/cli/utilities/find-up');
```

Reply:
It was a from scratch re-write, not a copy.

But yeah, we could do that, with the caveat that it is not documented as a public api, so could potentially break in the future. (On the other hand custom builders were not documented as a public API before angular 8, and that was not a considered a problem for this package for angular 7.)

I'm perfectly happy to switch to that if that's what you want to do.

Reply:
I vote we switch over to `@angular/cli/utilities/find-up`